rllib/env/multi_agent_episode.py

```diff
@@ -0,0 +1,718 @@
+
+        if initial_render_image is not None:
+            self.render_images.append(initial_render_image)
+
+        # Note, all agents will have an initial observation.
+        for agent_id in initial_observation.keys():
+            # Add initial timestep for each agent to the timestep mapping.
+            self.global_t_to_local_t[agent_id].append(self.t)
+
+            # Add initial observations to the agent's episode.
+            self.agent_episodes[agent_id].add_initial_observation(
+                # Note, initial observation has to be provided.
+                initial_observation=initial_observation[agent_id],
+                initial_state=None
+                if initial_state is None
+                else initial_state[agent_id],
+            )
+
+    def add_timestep(
+        self,
+        observation: MultiAgentDict,
+        action: MultiAgentDict,
+        reward: MultiAgentDict,
+        *,
+        state: Optional[MultiAgentDict] = None,
+        is_terminated: Optional[bool] = None,
+        is_truncated: Optional[bool] = None,
+        render_image: Optional[np.ndarray] = None,
+    ) -> None:
+        # Cannot add data to an already done episode.
+        assert not self.is_done
+
+        # Environment step.
+        self.t += 1
+
+        # TODO (sven, simon): Wilol there still be an `__all__` that is
+        # terminated or truncated?
+        self.is_terminated = (
+            False if is_terminated is None else is_terminated["__all__"]
+        )
+        self.is_truncated = False if is_truncated is None else is_truncated["__all__"]
+
+        # Add data to agent episodes.
+        for agent_id in observation.keys():
+            # If the agent stepped we need to keep track in the timestep mapping.
+            self.global_t_to_local_t[agent_id].append(self.t)
+
+            if render_image is not None:
+                self.render_images.append(render_image)
+
+            self.agent_episodes[agent_id].add_timestep(
+                observation[agent_id],
+                action[agent_id],
+                reward[agent_id],
+                state=None if state is None else state[agent_id],
+                is_terminated=None
+                if is_terminated is None
+                else is_terminated[agent_id],
+                is_truncated=None if is_truncated is None else is_truncated[agent_id],
+                render_image=None if render_image is None else render_image[agent_id],
+            )
+
+    @property
+    def is_done(self):
+        """Whether the episode is actually done (terminated or truncated).
+
+        A done episode cannot be continued via `self.add_timestep()` or being
+        concatenated on its right-side with another episode chunk or being
+        succeeded via `self.create_successor()`.
+
+        Note that in a multi-agent environment this does not necessarily
+        correspond to single agents having terminated or being truncated.
+
```

Comment:
Let's explain this here a little (what we discussed on slack):

```
For MultiAgentEpisode: self.is_terminated should be True, if all agents are terminated and self.is_truncated should be True if all agents are truncated.
If only one or more agents (but not all!) are terminated or truncated, the MultiAgentEpisode.is_terminated/is_truncated should both be False. The information about single agents' terminated/truncated states can always be retrieved from the SingleAgentEpisodes inside the MultiAgent one.
If all agents are either terminated or truncated (but in a mixed fashion: some agents terminated, some truncated): This is currently undefined and could potentially be a problem (if a user really implemented such a multi-agent env that behaves this way). My guess is that we should probably then set both is_terminated AND is_truncated in the MultiAgentEpisode to True. Question here: Does this have practical relevance?
```

Reply:
I think it has for postprocessing: If truncated, the `GAE` uses for example bootstrapping, while if `terminated` the value of $o_T$ is set to `0.0`.